engine/classes/Elgg/Http/ResponseFactory.php

```diff
@@ -0,0 +1,334 @@
+<?php
+
+namespace Elgg\Http;
+
+use Elgg\Ajax\Service as AjaxService;
+use Elgg\PluginHooksService;
+use SecurityException;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Response;
+
+/**
+ * WARNING: API IN FLUX. DO NOT USE DIRECTLY.
+ *
+ * @since 2.2.0
+ * @access private
+ */
+class ResponseFactory {
+
+	/**
+	 * @var Request
+	 */
+	private $request;
+
+	/**
+	 * @var AjaxService
+	 */
+	private $ajax;
+
+	/**
+	 * @var PluginHooksService
+	 */
+	private $hooks;
+
+	/**
+	 * Constructor
+	 * 
+	 * @param Request            $request HTTP request
+	 * @param PluginHooksService $hooks   Plugin hooks service
+	 * @param AjaxService        $ajax    AJAX service
+	 */
+	public function __construct(Request $request, PluginHooksService $hooks, AjaxService $ajax) {
+		$this->request = $request;
+		$this->ajax = $ajax;
+		$this->hooks = $hooks;
+	}
+
+	/**
+	 * Creates an HTTP response
+	 *
+	 * @param string  $content The response content
+	 * @param integer $status  The response status code
+	 * @param array   $headers An array of response headers
+	 * @return Response
+	 */
+	public function prepareResponse($content = '', $status = 200, array $headers = array()) {
+		$response = new Response($content, $status, $headers);
+		$response->prepare($this->request);
+		return $response;
+	}
+
+	/**
+	 * Creates a redirect response
+	 *
+	 * @param string  $url     URL to redirect to
+	 * @param integer $status  The status code (302 by default)
+	 * @param array   $headers An array of response headers (Location is always set to the given URL)
+	 * @return RedirectResponse
+	 */
+	public function prepareRedirectResponse($url, $status = 302, array $headers = array()) {
+		$response = new RedirectResponse($url, $status, $headers);
+		$response->prepare($this->request);
+		return $response;
+	}
+
+	/**
+	 * Send HTTP response
+	 *
+	 * @param ResponseBuilder $response An instanceof an ErrorResponse, OkResponse or RedirectResponse
+	 * @return Response
+	 */
+	public function respond(ResponseBuilder $response) {
+
+		$response_type = $this->parseContext();
+		$response = $this->hooks->trigger('response', $response_type, $response, $response);
+
+		$is_xhr = $this->request->isXmlHttpRequest();
+
+		$is_action = false;
+		if (0 === strpos($response_type, 'action:')) {
+			$is_action = true;
+		}
+
+		if ($is_action && $response->getForwardURL() === null) {
+			// actions must always set a redirect url
+			$response->setForwardURL(REFERRER);
+		}
+
+		if ($response->getForwardURL() !== null && !$is_xhr && $response->isOK()) {
+			// non-xhr requests should issue a forward if redirect url is set
+			$response->setStatusCode(ELGG_HTTP_FOUND);
+		}
+
+		if ($is_xhr && $is_action && !$this->ajax->isAjax2Request()) {
+			// xhr actions using legacy ajax API should return 200 with wrapped data
+			$response->setStatusCode(ELGG_HTTP_OK);
```

Comment:
XHR calls to static resources `ajax/view` and `ajax/form` will get their 200 code with the response. Others are likely ending with `forward()` which will trigger ajax forward hook, which will wrap the response.
XHR errors will be handled as before (via elgg/Ajax, or with an error header for legacy requests)